File: src/Day05.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;

import static java.util.Collections.emptySet;
import static java.util.stream.Collectors.*;
import static java.util.stream.Gatherers.windowSliding;

public class Day05 {

  private static List<List<Integer>> parsePageLists(List<String> input) {
    return input.stream()
        .filter(line -> line.contains(","))
        .map(line -> Arrays
            .stream(line.split(","))
            .map(Integer::parseInt)
            .toList())
        .toList();
  }

  @SuppressWarnings("ComparatorMethodParameterNotUsed")
  private static Comparator<Integer> comparator(List<String> input) {
    Map<Integer, Set<Integer>> precedences = input.stream()
        .filter(line -> line.contains("|"))
        .map(line -> Arrays
            .stream(line.split("\\|"))
            .map(Integer::parseInt)
            .toArray(Integer[]::new))
        .collect(groupingBy(
            arr -> arr[0],
            mapping(arr -> arr[1], toSet())
        ));

    return (o1, o2) -> precedences.getOrDefault(o1, emptySet()).contains(o2) ? -1 : 1;
  }

  private static int part1(List<String> input) {
    Comparator<Integer> comparator = comparator(input);
    return parsePageLists(input)
        .stream()
        .filter(list -> list
            .stream()
            .gather(windowSliding(2))
            .noneMatch(i -> comparator.compare(i.getFirst(), i.get(1)) >= 0))
        .mapToInt(list -> list.get(list.size() / 2))
        .sum();
  }

  private static int part2(List<String> input) {
    Comparator<Integer> comparator = comparator(input);
    return parsePageLists(input).stream()
        .filter(list -> list
            .stream()
            .gather(windowSliding(2))
            .anyMatch(i -> comparator.compare(i.getFirst(), i.get(1)) >= 0))
        .map(list -> list.stream().sorted(comparator).toList())
        .mapToInt(list -> list.get(list.size() / 2))
        .sum();
  }

  private static List<String> readInput(String filename) throws IOException {
    return Files.lines(Path.of(filename)).toList();
  }

  public static void main(String[] args) throws IOException {
    if (args.length == 0) {
      List<String> testInput = readInput("src/05t.txt");
      assert part1(testInput) == 143 : "Expected 143 but got " + part1(testInput);
      List<String> input = readInput("src/05.txt");
      System.out.println(part1(input));
      assert part2(testInput) == 123 : "Expected 123 but got " + part2(testInput);
      System.out.println(part2(input));
    } else {
      List<String> input = Files.readAllLines(Path.of(args[0]));
      System.out.println(part1(input));
      System.out.println(part2(input));
    }
  }
}

File: src/Day05.kt
import kotlin.io.path.Path
import kotlin.io.path.readText


fun main(args: Array<String>) {

    fun parsePageLists(input: List<String>): List<List<Int>> {
        return input.filter { it.contains(",") }.map { it.split(",").map { it.toInt() } }
    }

    fun comparator(input: List<String>): Comparator<Int> {
        val precedences = input
            .filter { it.contains("|") }
            .map { it.split("|").map { it.toInt() } }
            .groupBy({ it[0] }, { it[1] })
            .mapValues { it.value.toSet() }
        return Comparator { o1, o2 -> if (precedences[o1]?.contains(o2) == true) -1 else 1 }
    }

    fun part1(input: List<String>): Int {
        val comparator = comparator(input)
        return parsePageLists(input)
            .filter { it.windowed(2).all { (a, b) -> comparator.compare(a, b) == -1 } }
            .sumOf { it[it.size / 2] }
    }

    fun part2(input: List<String>): Int {
        val comparator = comparator(input)
        return parsePageLists(input)
            .filter { it.windowed(2).find { (a, b) -> comparator.compare(a, b) == 1 } != null }
            .map { it.sortedWith(comparator) }
            .sumOf { it[it.size / 2] }
    }


    // Or read a large test input from the `src/Day01_test.txt` file:
    if (args.isEmpty()) {
        val testInput = readInput("05t")
        check(part1(testInput) == 143) { "Expected 143 but got ${part1(testInput)}" }

        // Read the input from the `src/Day01.txt` file.
        val input = readInput("05")
        part1(input).println()
        check(part2(testInput) == 123) { "Expected 123 but got ${part2(testInput)}" }
        part2(input).println()
    } else {
        val input = Path(args[0]).readText().trim().lines()
        part1(input).println()
        part2(input).println()
    }
}


